File: Backtracking/140_WordBreakII.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        res = []
        dict_set = set(wordDict)
        n = len(s)

        def backtrack(left, words):
            if left == n:
                res.append(" ".join(words))
                return
            for word in dict_set:
                if s[left:].startswith(word):
                    words.append(word)
                    backtrack(left + len(word), words)
                    words.pop()

        backtrack(0, [])
        return res


class Solution2:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        res = []
        word_set = set(wordDict)
        n = len(s)

        def backtrack(start, cur):
            if start == n:
                res.append(" ".join(cur))
                return
            for end in range(start + 1, n + 1):
                word = s[start: end]
                if word in word_set:
                    cur.append(word)
                    backtrack(end, cur)
                    cur.pop()

        backtrack(0, [])
        return res

File: Java Solutions/140_Word Break II.java
class Solution {
    private String s;
    private Set<String> wordSet;
    private List<String> res = new ArrayList<>();

    public List<String> wordBreak(String s, List<String> wordDict) {
        this.s = s;
        this.wordSet = new HashSet<>(wordDict);
        backtrack(0, new StringBuilder());
        return res;
    }

    private void backtrack(int start, StringBuilder curSentence) {
        if (start == s.length()) {
            res.add(curSentence.toString().trim());
            return;
        }

        for (int end = start + 1; end <= s.length(); end++) {
            String word = s.substring(start, end);
            if (wordSet.contains(word)) {
                int curLength = curSentence.length();
                curSentence.append(word).append(" ");
                backtrack(end, curSentence);
                curSentence.setLength(curLength);
            }
        }
    }
}